<a href="https://colab.research.google.com/github/zzzer0-wav/myDTA_2026/blob/main/classmade/PANDAS_tasks_beginners_DTA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Практичні завдання з pandas для початківців

**Мета:** Набути базових навичок роботи з бібліотекою pandas для аналізу табличних даних.

**Інструкції:**
- Виконуйте завдання послідовно — кожне наступне спирається на попередні.
- Після виконання кожного завдання напишіть короткий висновок у комірці `**Висновок:**`.
- Не змінюйте код у комірках з підготовкою даних.

**Набір даних:** Умовна таблиця продажів інтернет-магазину за 2023 рік.

---
## Підготовка: Генерація тестових даних

Запустіть цю комірку перед початком роботи. Не змінюйте код.

In [2]:
import pandas as pd
import numpy as np

np.random.seed(42)
n = 200

categories = ['Електроніка', 'Одяг', 'Книги', 'Продукти', 'Спорт']
cities = ['Київ', 'Харків', 'Одеса', 'Дніпро', 'Львів']
payment_methods = ['Картка', 'Готівка', 'Онлайн']

df = pd.DataFrame({
    'order_id':        range(1001, 1001 + n),
    'date':            pd.date_range('2023-01-01', periods=n, freq='D')[:n],
    'category':        np.random.choice(categories, n),
    'product':         ['Товар_' + str(i) for i in np.random.randint(1, 51, n)],
    'city':            np.random.choice(cities, n),
    'quantity':        np.random.randint(1, 10, n),
    'price':           np.round(np.random.uniform(50, 5000, n), 2),
    'payment_method':  np.random.choice(payment_methods, n),
    'rating':          np.random.choice([1, 2, 3, 4, 5, np.nan], n, p=[0.05, 0.1, 0.15, 0.3, 0.3, 0.1]),
    'is_returned':     np.random.choice([True, False], n, p=[0.1, 0.9]),
})

# Додаємо кілька пропущених значень
df.loc[np.random.choice(df.index, 10, replace=False), 'city'] = np.nan
df.loc[np.random.choice(df.index, 5, replace=False), 'price'] = np.nan

# Розраховуємо загальну суму
df['total'] = df['quantity'] * df['price']

print('Дані успішно згенеровано. Рядків:', len(df))
print('Колонки:', list(df.columns))

Дані успішно згенеровано. Рядків: 200
Колонки: ['order_id', 'date', 'category', 'product', 'city', 'quantity', 'price', 'payment_method', 'rating', 'is_returned', 'total']


---
## Завдання 1 (Легке). Перший погляд на дані

Виведіть перші 10 рядків датафрейму `df` та отримайте загальну інформацію про нього (кількість рядків, колонок, типи даних) за допомогою методів `head()` та `info()`.

In [3]:
# Ваш код тут
display(df.head(10))
display(df.info())
display(df.shape)
display(df.isna().sum())

,order_id,date,category,product,city,quantity,price,payment_method,rating,is_returned,total
0,1001,2023-01-01,Продукти,Товар_34,Київ,1,963.38,Картка,5.0,False,963.38
1,1002,2023-01-02,Спорт,Товар_6,Львів,6,1086.28,Картка,4.0,False,6517.68
2,1003,2023-01-03,Книги,Товар_22,Дніпро,4,1883.84,Готівка,5.0,False,7535.36
3,1004,2023-01-04,Спорт,Товар_11,Дніпро,1,2448.39,Картка,5.0,False,2448.39
4,1005,2023-01-05,Спорт,Товар_48,Дніпро,7,3110.36,Онлайн,3.0,False,21772.52
5,1006,2023-01-06,Одяг,Товар_16,Львів,9,1876.12,Онлайн,NaN,False,16885.08
6,1007,2023-01-07,Книги,Товар_33,Дніпро,4,2339.55,Картка,4.0,False,9358.20
7,1008,2023-01-08,Книги,Товар_9,Львів,4,3749.98,Готівка,4.0,False,14999.92
8,1009,2023-01-09,Книги,Товар_6,Дніпро,6,231.58,Онлайн,4.0,False,1389.48
9,1010,2023-01-10,Спорт,Товар_16,Одеса,3,1299.56,Готівка,1.0,False,3898.68


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   order_id        200 non-null    int64         
 1   date            200 non-null    datetime64[ns]
 2   category        200 non-null    object        
 3   product         200 non-null    object        
 4   city            190 non-null    object        
 5   quantity        200 non-null    int64         
 6   price           195 non-null    float64       
 7   payment_method  200 non-null    object        
 8   rating          177 non-null    float64       
 9   is_returned     200 non-null    bool          
 10  total           195 non-null    float64       
dtypes: bool(1), datetime64[ns](1), float64(3), int64(2), object(4)
memory usage: 15.9+ KB


None

(200, 11)

,0
order_id,0
date,0
category,0
product,0
city,10
quantity,0
price,5
payment_method,0
rating,23
is_returned,0


**Висновок:** *(Напишіть, скільки рядків і колонок у датафреймі, які типи даних зустрічаються, чи є пропущені значення)*

___________________

Датафрейм має у собі 200 рядків та 11 колонок.

Зустрічаються типи даних цілих чисел, дробових чисел, бульові типи даних, дата та час, та змішані (строки, числа, none).

Також у датафреймі зустрічаються пусті значення у колонках city (10), price (5), rating (23) та total (5).

---
## Завдання 2 (Легке). Описова статистика

Використайте метод `describe()` для отримання описової статистики числових колонок. Окремо виведіть мінімум, максимум і середнє значення колонки `price`.

In [4]:
# Ваш код тут
print(df.describe())
print()
print('Максимальний price: ',df['price'].max(),'$\n')
print('Мілімальний price: ',df['price'].min(),'$\n')
print('Середнє значення price: ',round(df['price'].mean(),2),'$\n')

          order_id                 date    quantity        price      rating  \
count   200.000000                  200  200.000000   195.000000  177.000000   
mean   1100.500000  2023-04-10 12:00:00    4.700000  2500.579641    3.813559   
min    1001.000000  2023-01-01 00:00:00    1.000000   139.650000    1.000000   
25%    1050.750000  2023-02-19 18:00:00    2.000000  1288.215000    3.000000   
50%    1100.500000  2023-04-10 12:00:00    5.000000  2517.290000    4.000000   
75%    1150.250000  2023-05-30 06:00:00    7.000000  3706.975000    5.000000   
max    1200.000000  2023-07-19 00:00:00    9.000000  4950.300000    5.000000   
std      57.879185                  NaN    2.618066  1380.066748    1.169808   

              total  
count    195.000000  
mean   11629.097590  
min      149.350000  
25%     3893.955000  
50%     8746.840000  
75%    16922.420000  
max    44324.640000  
std     9968.293697  

Максимальний price:  4950.3 $

Мілімальний price:  139.65 $

Середнє значення pr

**Висновок:** *(Опишіть діапазон цін, середнє значення, чи є суттєвий розкид)*
___
Ціни у вибірці знаходяться у діапазоні від 139.65$ по 4950.3$, середнє значення становить 2500.58$, що близьке до медіани - 2517.29$, тому розподіл цін є відносно збалансованим.

Так, розкид значень доволі суттєвий, (Стандартне відхилення ≈ 1380$), що вказує на наявність як дешевих таки дорогих товарів у вибірці.

---
## Завдання 3 (Легке). Пропущені значення

Знайдіть кількість пропущених значень у кожній колонці за допомогою `isnull().sum()`. Визначте, який відсоток рядків має пропущені значення хоча б в одній колонці (`dropna()` та `len()`).

In [5]:
# Ваш код тут
print(df.isna().sum())
print(f'\nКількість NaN у датафреймі: {df.isnull().sum().sum()}')
print(f'\nКількість строк з NaN: {df.isna().any(axis=1).sum()}')
rows_with_na = len(df) - len(df.dropna())
percent = rows_with_na / len(df) * 100

print(f'\n{percent}% рядків містять хоча б одне пропущене значення')

order_id           0
date               0
category           0
product            0
city              10
quantity           0
price              5
payment_method     0
rating            23
is_returned        0
total              5
dtype: int64

Кількість NaN у датафреймі: 43

Кількість строк з NaN: 37

18.5% рядків містять хоча б одне пропущене значення


**Висновок:** *(У яких колонках є пропуски, наскільки це критично для аналізу)*
___
Пропуски є у колонках city, price, rating та total.


Пропуски у price & total - є критичними, для підрахунку загального прибутку та виторга. Їх мало, по 5 шт., можна заповнити медіаною або видалити.

Відсутні значення у city - це може суттєво змінити результати аналізу по регіонах а відсутній рейгинг не є критичним для фінансового аналізу, але треба його враховувати якщо хочемо аналізувати задоволеність клієнтів.

---
## Завдання 4 (Легке). Відбір колонок і рядків

Створіть новий датафрейм `df_subset`, що містить лише колонки `order_id`, `category`, `city`, `total`. Потім виведіть рядки з індексами від 10 до 20 включно за допомогою `.iloc[]`.

In [6]:
# Ваш код тут
df_subset = df[['order_id', 'category', 'city', 'total']]

df_subset.iloc[10:21]

,order_id,category,city,total
10,1011,Продукти,Дніпро,21486.48
11,1012,Книги,Львів,31368.89
12,1013,Спорт,Харків,7748.40
13,1014,Одяг,Дніпро,18787.72
14,1015,Продукти,NaN,1741.50
15,1016,Одяг,Одеса,4529.38
16,1017,Продукти,Київ,10745.84
17,1018,Спорт,Одеса,10001.84
18,1019,Електроніка,Дніпро,12444.75
19,1020,Продукти,Харків,13422.92


**Висновок:** *(Поясніть різницю між вибором колонок через список і через `iloc`)*
___

.loc - вибірка по індексу | назві колонки df.loc[0:3, 'items']

.iloc - вибірка по індексу | номеру стовбчика df.iloc[0:3, 6:8]     (0-2 рядок, 6-7 стовбчик)

---
## Завдання 5 (Легке). Фільтрація даних

Відфільтруйте датафрейм і отримайте:
1. Всі замовлення з категорії `'Електроніка'`.
2. Всі замовлення, де `price` перевищує 2000.
3. Поверненi замовлення (`is_returned == True`) з міста `'Київ'`.

Виведіть кількість рядків для кожного фільтру.

In [7]:
# Ваш код тут
electronics = df[df['category']=='Електроніка']
print('Всі замовлення з категорії \'Електроніка\':')
display(electronics.head())
print(f'Кількість рядків: {electronics.shape[0]}')

big_price = df[df['price']>2000]
print(f'\nВсі замовлення, де price перевищує 2000:')
display(big_price.head())
print(f'Кількість рядків: {big_price.shape[0]}')

mask = df[(df['city']=='Київ') & (df['is_returned']==True)]
print(f'\nПоверненi замовлення (is_returned == True) з міста \'Київ\':')
display(mask.head())
print(f'Кількість рядків: {mask.shape[0]}')

total = df.shape[0]

print(f'\nЕлектроніка: {electronics.shape[0]} з {total} ({electronics.shape[0]/total*100:.1f}%)')
print(f'Ціна > 2000: {big_price.shape[0]} з {total} ({big_price.shape[0]/total*100:.1f}%)')
print(f'Повернення з Києва: {mask.shape[0]} з {total} ({mask.shape[0]/total*100:.1f}%)')

Всі замовлення з категорії 'Електроніка':


,order_id,date,category,product,city,quantity,price,payment_method,rating,is_returned,total
18,1019,2023-01-19,Електроніка,Товар_20,Дніпро,9,1382.75,Онлайн,5.0,False,12444.75
23,1024,2023-01-24,Електроніка,Товар_40,Львів,1,1671.11,Онлайн,NaN,False,1671.11
24,1025,2023-01-25,Електроніка,Товар_39,Київ,9,642.82,Готівка,5.0,False,5785.38
33,1034,2023-02-03,Електроніка,Товар_31,Львів,5,3739.93,Готівка,5.0,False,18699.65
38,1039,2023-02-08,Електроніка,Товар_16,Львів,9,3247.29,Картка,5.0,False,29225.61


Кількість рядків: 43

Всі замовлення, де price перевищує 2000:


,order_id,date,category,product,city,quantity,price,payment_method,rating,is_returned,total
3,1004,2023-01-04,Спорт,Товар_11,Дніпро,1,2448.39,Картка,5.0,False,2448.39
4,1005,2023-01-05,Спорт,Товар_48,Дніпро,7,3110.36,Онлайн,3.0,False,21772.52
6,1007,2023-01-07,Книги,Товар_33,Дніпро,4,2339.55,Картка,4.0,False,9358.20
7,1008,2023-01-08,Книги,Товар_9,Львів,4,3749.98,Готівка,4.0,False,14999.92
10,1011,2023-01-11,Продукти,Товар_29,Дніпро,6,3581.08,Готівка,3.0,False,21486.48


Кількість рядків: 116

Поверненi замовлення (is_returned == True) з міста 'Київ':


,order_id,date,category,product,city,quantity,price,payment_method,rating,is_returned,total
68,1069,2023-03-10,Продукти,Товар_48,Київ,2,3858.99,Готівка,3.0,True,7717.98
110,1111,2023-04-21,Електроніка,Товар_47,Київ,3,863.12,Готівка,4.0,True,2589.36
135,1136,2023-05-16,Книги,Товар_21,Київ,2,3980.42,Готівка,5.0,True,7960.84
140,1141,2023-05-21,Електроніка,Товар_18,Київ,3,800.79,Онлайн,NaN,True,2402.37
141,1142,2023-05-22,Спорт,Товар_25,Київ,2,1593.02,Картка,3.0,True,3186.04


Кількість рядків: 7

Електроніка: 43 з 200 (21.5%)
Ціна > 2000: 116 з 200 (58.0%)
Повернення з Києва: 7 з 200 (3.5%)


In [8]:
df['price'].mean()

np.float64(2500.579641025641)

**Висновок:** *(Яка частка замовлень відповідає кожному критерію? Що можна сказати про повернення з Києва?)*
___

Замовлення категорії 'Електроніка' складають 21.5% від всіх замовлень.

Замовлень з чеком більше ніж 2000 - цілих 58%, середній чек складає 2500.58, це означає, що більшість покупців витрачають суму, вищу за 2000.

Повернень з міста Київ - 7 з 200 : 3.5%. Це відносно невелика частка, тому Київ не є проблемним регіоном з точки зору повернень.

---
## Завдання 6 (Середнє). Сортування та ранжування

1. Відсортуйте датафрейм за колонкою `total` у спадному порядку та виведіть топ-10 найбільших замовлень.
2. Додайте нову колонку `rank`, що містить порядковий номер замовлення за розміром суми (від найбільшого до найменшого) за допомогою `.rank()`.

In [9]:
# Ваш код тут:
df.sort_values('total', ascending=False).head(10)


,order_id,date,category,product,city,quantity,price,payment_method,rating,is_returned,total
190,1191,2023-07-10,Одяг,Товар_22,Київ,9,4924.96,Готівка,4.0,True,44324.64
58,1059,2023-02-28,Продукти,Товар_33,Харків,9,4848.88,Картка,5.0,False,43639.92
186,1187,2023-07-06,Спорт,Товар_34,Київ,9,4675.12,Готівка,5.0,False,42076.08
55,1056,2023-02-25,Електроніка,Товар_3,Дніпро,8,4912.77,Картка,5.0,False,39302.16
169,1170,2023-06-19,Продукти,Товар_50,Київ,8,4739.05,Готівка,5.0,False,37912.40
191,1192,2023-07-11,Спорт,Товар_25,Харків,9,4202.55,Готівка,4.0,True,37822.95
193,1194,2023-07-13,Одяг,Товар_22,NaN,8,4608.17,Картка,4.0,False,36865.36
51,1052,2023-02-21,Продукти,Товар_35,Дніпро,8,4488.99,Онлайн,4.0,False,35911.92
25,1026,2023-01-26,Книги,Товар_18,Дніпро,8,4458.11,Картка,3.0,False,35664.88
143,1144,2023-05-24,Продукти,Товар_31,Київ,9,3732.53,Готівка,5.0,False,33592.77


In [10]:
df['rank'] = df['total'].rank(ascending=False)
df.head()

,order_id,date,category,product,city,quantity,price,payment_method,rating,is_returned,total,rank
0,1001,2023-01-01,Продукти,Товар_34,Київ,1,963.38,Картка,5.0,False,963.38,188.0
1,1002,2023-01-02,Спорт,Товар_6,Львів,6,1086.28,Картка,4.0,False,6517.68,117.0
2,1003,2023-01-03,Книги,Товар_22,Дніпро,4,1883.84,Готівка,5.0,False,7535.36,109.0
3,1004,2023-01-04,Спорт,Товар_11,Дніпро,1,2448.39,Картка,5.0,False,2448.39,171.0
4,1005,2023-01-05,Спорт,Товар_48,Дніпро,7,3110.36,Онлайн,3.0,False,21772.52,28.0


**Висновок:** *(Яка максимальна сума замовлення? З яких категорій найбільші замовлення?)*

Максимальна сума замовлення - 443424.64 грн категорії "Одяг", купили 9шт по 4924.96 грн.

Судячи з топу максимальних замовлень по сумі, частіше зустрічається категорія "Продукти" (4 з 10), далі йдуть "Спорт" та "Одяг" по 2 заповлення.

Можна зробити висновок, що веліки суми чеків частіше формуються за рахунок великої кількості одиниць товару, а не лише високої ціни.

---
## Завдання 7 (Середнє). Групування та агрегація

За допомогою `groupby()` розрахуйте для кожної категорії товарів:
- загальний обсяг продажів (сума `total`),
- кількість замовлень,
- середню ціну (`price`).

Відсортуйте результат за загальним обсягом продажів.

In [11]:
# Ваш код тут
result = df.groupby('category').agg(
    total_sum=('total', 'sum'),
    orders_count=('order_id', 'count'),
    avg_price=('price', 'mean')
).sort_values('total_sum', ascending=False).round(2)

result



,total_sum,orders_count,avg_price
category,,,
Продукти,583761.68,49,2513.98
Одяг,474135.61,35,2707.91
Електроніка,428129.79,43,2334.47
Спорт,400802.56,36,2494.21
Книги,380844.39,37,2491.88


**Висновок:** *(Яка категорія приносить найбільший дохід? Чи збігається це з кількістю замовлень?)*

Найбільший дохід приносить категорія "Продукти" - 583761.68 грн, також вона має найбільшу кількість замовлень (49), та середня сума чеку - 2513.98 грн.
Дохід та кількість збігаються.

Проте "Одяг" має менше замовлень (35), але вищий середній чек (2707.91), що свідчить про те, що покупці витрачають більше за одне замовлення.


---
## Завдання 8 (Середнє). Обробка пропущених значень

1. Заповніть пропущені значення в колонці `price` медіаною цін відповідної категорії (використайте `groupby().transform('median')`).
2. Заповніть пропущені значення в колонці `city` рядком `'Невідомо'`.
3. Пропущені значення в `rating` залиште без змін — поясніть чому у висновку.

Перевірте, що пропусків у `price` та `city` більше немає.

In [29]:
# Ваш код тут
median_by_category = df.groupby('category')['price'].transform('median')
df['price'] = df['price'].fillna(median_by_category)
df['city'] = df['city'].fillna('Невідомо')

print(df['price'].isna().sum())
print(df['city'].isna().sum())

0
0


**Висновок:** *(Чому медіана краща за середнє для заповнення цін? Чому пропуски в rating можна залишити?)*

Медіана краще за середнє для заповнення цін, тому що у датасеті можуть бути "викиди"/аномальні значення, завдяки яким середнє значення буде сильно завищено або занижено. Медіана допомагає уникнути таких підрахункових помилок.

Рейтинг був залишений без змін, тому що це суб'єктивна оцінка покупця, ми не можемо замінити значення медіаною або середнім, тому зо це буде некоректно.

---
## Завдання 9 (Середнє). Додавання обчислюваних колонок

Додайте до датафрейму нові колонки:
1. `month` — місяць замовлення (з колонки `date`).
2. `price_category` — категорія ціни: `'Низька'` (до 500), `'Середня'` (500–2000), `'Висока'` (понад 2000). Використайте `pd.cut()` або `np.select()`.
3. `discount_total` — сума зі знижкою 15%.

In [13]:
# Ваш код тут


**Висновок:** *(Який розподіл цінових категорій? Як метод pd.cut спрощує категоризацію?)*

---
## Завдання 10 (Середнє). Зведена таблиця (pivot table)

Побудуйте зведену таблицю за допомогою `pd.pivot_table()`, де:
- рядки — `city`,
- колонки — `category`,
- значення — сума `total`,
- агрегація — `sum`.

Додайте підсумки по рядках і колонках (`margins=True`).

In [14]:
# Ваш код тут


**Висновок:** *(Яке місто і яка категорія дають найбільший сумарний дохід? Що показують margins?)*

---
## Завдання 11 (Середнє). Підрахунок унікальних значень і частот

1. Підрахуйте кількість унікальних товарів (`product`) у кожній категорії.
2. Виведіть топ-5 найпопулярніших товарів за кількістю замовлень за допомогою `value_counts()`.
3. Визначте частку кожного способу оплати у відсотках (`normalize=True`).

In [15]:
# Ваш код тут


**Висновок:** *(Який спосіб оплати найпопулярніший? Чи рівномірний розподіл між способами?)*

---
## Завдання 12 (Середнє). Аналіз повернень

Розрахуйте рівень повернень (частку повернених замовлень) для кожної категорії та кожного міста. Визначте:
- категорію з найвищим рівнем повернень,
- місто з найнижчим рівнем повернень.

Підказка: використайте `groupby()['is_returned'].mean()`.

In [16]:
# Ваш код тут


**Висновок:** *(Яка категорія найчастіше повертається? Про що це може свідчити?)*

---
## Завдання 13 (Підвищена складність). Часовий аналіз

1. Встановіть колонку `date` як індекс датафрейму.
2. Розрахуйте щомісячний обсяг продажів (ресемплінг `resample('ME')` або `resample('M')`, sum по `total`).
3. Знайдіть найкращий і найгірший місяці за продажами.
4. Поверніть звичайний числовий індекс (`reset_index()`).

In [17]:
# Ваш код тут


**Висновок:** *(Який місяць виявився найуспішнішим? Чи є сезонність у даних?)*

---
## Завдання 14 (Підвищена складність). Накопичувальні показники

1. Відсортуйте датафрейм за датою.
2. Додайте колонку `cumulative_total` — накопичувальний обсяг продажів з початку року (`cumsum()`).
3. Визначте, в який день було досягнуто 50% від загального річного обсягу продажів.

In [18]:
# Ваш код тут


**Висновок:** *(Коли приблизно була зароблена половина річного доходу? Що це означає для бізнесу?)*

---
## Завдання 15 (Підвищена складність). apply() та lambda

1. Використайте `apply()` з lambda-функцією для створення колонки `total_with_tax`, де ПДВ становить 20%, але лише для замовлень на суму понад 1000 (решта — без змін).
2. Напишіть функцію `classify_rating(r)`, що повертає `'Поганий'` (1–2), `'Нейтральний'` (3), `'Хороший'` (4–5), `'Без оцінки'` (NaN). Застосуйте її до колонки `rating`.

In [19]:
# Ваш код тут


**Висновок:** *(Яка частка замовлень має хороший рейтинг? Коли apply кращий за vectorized-операції?)*

---
## Завдання 16 (Підвищена складність). Злиття датафреймів (merge)

Створіть додатковий датафрейм `df_discounts` з колонками `category` та `discount_pct` (різні знижки для кожної категорії). Об'єднайте його з основним датафреймом за допомогою `pd.merge()` та розрахуйте `final_total` — суму після застосування відповідної знижки.

```python
df_discounts = pd.DataFrame({
    'category':     ['Електроніка', 'Одяг', 'Книги', 'Продукти', 'Спорт'],
    'discount_pct': [0.10, 0.20, 0.05, 0.00, 0.15]
})
```

In [20]:
df_discounts = pd.DataFrame({
    'category':     ['Електроніка', 'Одяг', 'Книги', 'Продукти', 'Спорт'],
    'discount_pct': [0.10, 0.20, 0.05, 0.00, 0.15]
})

# Ваш код тут


**Висновок:** *(Яка категорія отримала найбільшу сумарну знижку? Коли варто використовувати merge замість map?)*

---
## Завдання 17 (Підвищена складність). Групові трансформації

1. Додайте колонку `city_avg_total` — середня сума замовлення по місту (для кожного рядка підставте середнє по його місту), використайте `groupby().transform('mean')`.
2. Додайте колонку `above_city_avg` — булева колонка, що показує, чи перевищує замовлення середнє по місту.
3. Підрахуйте частку таких замовлень для кожного міста.

In [21]:
# Ваш код тут


**Висновок:** *(Чим transform відрізняється від звичайного groupby? У якому місті найбільша частка великих замовлень?)*

---
## Завдання 18 (Складне). Ковзне середнє та аномалії

1. Згрупуйте дані за місяцем та розрахуйте щоденну виручку.
2. Розрахуйте 7-денне ковзне середнє виручки (`rolling(7).mean()`).
3. Визначте дні-аномалії: дні, коли виручка відхилилася від ковзного середнього більш ніж на 2 стандартних відхилення.

In [22]:
# Ваш код тут


**Висновок:** *(Скільки аномальних днів виявлено? Що може бути причиною різких відхилень виручки?)*

---
## Завдання 19 (Складне). Когортний аналіз

Виконайте спрощений когортний аналіз за місяцем першого замовлення:
1. Для кожного товару (`product`) знайдіть місяць першого продажу — це його «когорта».
2. Об'єднайте цю інформацію з основним датафреймом.
3. Побудуйте зведену таблицю: когорта (місяць першого продажу) проти місяця замовлення, значення — кількість замовлень.

Підказка: використайте `groupby('product')['date'].transform('min')` та `dt.to_period('M')`.

In [23]:
# Ваш код тут


**Висновок:** *(Що показує когортна таблиця? Чи продовжують товари з першої когорти продаватися в наступних місяцях?)*

---
## Завдання 20 (Складне). Підсумковий звіт

Сформуйте підсумковий звіт у вигляді одного датафрейму `summary_report`, що містить для кожної категорії:
- загальний дохід (`total_revenue`),
- кількість замовлень (`order_count`),
- середній чек (`avg_order_value`),
- частку доходу від загального (`revenue_share_pct`),
- рівень повернень у відсотках (`return_rate_pct`),
- середній рейтинг без NaN (`avg_rating`).

Відсортуйте за `total_revenue` спадно. Збережіть результат у файл `summary_report.csv`.

In [24]:
# Ваш код тут


**Висновок:** *(Яка категорія є лідером за доходом? Чи є зв'язок між рейтингом та рівнем повернень? Які рекомендації можна дати бізнесу на основі цього звіту?)*

---
## Загальні підсумки

Після виконання всіх завдань дайте відповідь на такі питання:

1. Які методи pandas виявилися для вас найбільш корисними?
2. Які труднощі виникли під час роботи з пропущеними значеннями?
3. Що можна додати до аналізу, щоб зробити його більш повним?

*(Ваші відповіді тут)*